### PR TITLE
Cognito authorisation with API gateway

### PR DESCRIPTION
**Pull request by Biplob14**
Fixes #
Implemented authorisation process in API gateway endpoint

Resolves #
Used when the changes in the pull request indirectly resolve another issue.

- [ ] Does not depend on other PRs
- [x] Depends on other PR being merged first: #162 

## What kind of change does this PR introduce (check at least one)

- [ ] Bugfix
- [ ] Improvement
- [x] New feature
- [ ] Refactor
- [ ] Infrastructure
- [ ] Other (please specify)

## How would you describe the changes in a changelog (1-2 lines)?

Implemented authorisation to access `API` endpoints. Without appropriate key user won't be access specific endpoints. 

## Related Issues/PRs (optional)

Link to the relevant issue(s) or provide a clear description of the problem or feature being addressed.

## How to test

- [ ] Needs regression testing (add QA needed label)

## Checklist

- [ ] I have tested these changes locally and they function as expected.
- [ ] I have added or updated necessary documentation to reflect the changes.
- [ ] I have followed the project's coding style and guidelines.
- [ ] I have added appropriate test coverage to ensure the changes are functioning correctly.
- [ ] I have checked for any potential conflicts with other branches or pull requests.
- [ ] I have updated necessary OpenAPI specs

## Reviewer Request (optional)

If you have any specific requests or suggestions for reviewers, mention them here.

## Screenshots (optional)

If applicable, provide any additional screenshots or GIFs that showcase the changes made in this pull request.

## Demo Link (optional)

If applicable, provide a link to a live demo or deployment where the changes can be reviewed.

## Known Issues (optional)

If there are any known issues or limitations with the changes made in this pull request, describe them here.

## Additional Notes (optional)

Add any additional notes or comments you think are relevant to this pull request.

## Deploy risk & follow up

Please assess risk & follow up on deployment of this PR. Elaborate your choice if selecting anything other than 'No risk'.

- [x] No risk
- [ ] Low
- [ ] High

### CODE REVIEW

File: iac/terraform/aws/openapi.tf [hunk 1]
```diff
@@ -12,8 +12,9 @@ data "template_file" "openapi_definition" {
   template = file(var.combined_openapi_file)
 
   vars = merge({
-    ENVIRONMENT = var.environment
+    ENVIRONMENT = var.environment
```
Comment (mahinuzzaman) [inline iac/terraform/aws/openapi.tf:15]: Please remove the extra space

File: iac/terraform/aws/cognito/cognito.tf [hunk 2]
```diff
@@ -56,6 +56,11 @@ resource "aws_cognito_user_pool" "user_pool" {
 
 }
 
+resource "aws_cognito_user_pool_domain" "name" {
+  domain = "bloodconnect"
```
Comment (sakibstark11) [inline iac/terraform/aws/cognito/cognito.tf:60]: why is the domain hardcoded?

Comment (Biplob14) [reply iac/terraform/aws/cognito/cognito.tf:60]: It's not the full domain name and It's used just one place. That's why I didn't declared another variable for this.

File: iac/terraform/aws/cognito/cognito.tf [hunk 3]
```diff
@@ -69,3 +74,17 @@ resource "aws_cognito_user_pool_client" "app_pool_client" {
...
+  name         = "user"
+  description  = "Standard user group"
+  precedence   = 1
+}
+
+resource "aws_cognito_user_group" "organization_group" {
+  user_pool_id = aws_cognito_user_pool.user_pool.id
+  name         = "organization"
+  description  = "Organization user group"
+  precedence   = 0
```
Comment (KhirulIslam) [inline iac/terraform/aws/cognito/cognito.tf:89]: do we need `precedence`? I don't see any usage

File: iac/terraform/aws/cognito/cognito.tf [hunk 4]
```diff
@@ -69,3 +74,17 @@ resource "aws_cognito_user_pool_client" "app_pool_client" {
   supported_identity_providers         = ["COGNITO"]
   explicit_auth_flows                  = ["ALLOW_USER_PASSWORD_AUTH", "ALLOW_REFRESH_TOKEN_AUTH", "ALLOW_USER_SRP_AUTH"]
 }
+
+resource "aws_cognito_user_group" "user_group" {
+  user_pool_id = aws_cognito_user_pool.user_pool.id
+  name         = "user"
+  description  = "Standard user group"
+  precedence   = 1
```
Comment (KhirulIslam) [inline iac/terraform/aws/cognito/cognito.tf:82]: same as the other

File: iac/terraform/aws/openapi.tf (hunk 1 above)
Comment (Biplob14) [reply iac/terraform/aws/openapi.tf]: Removed them

File: iac/terraform/aws/cognito/cognito.tf (hunk 2 above)
Comment (Biplob14) [reply iac/terraform/aws/cognito/cognito.tf]: changed the code. Used variable
https://github.com/CraftsmenLtd/BloodConnect/pull/163/files#diff-de0ee50e8abad6227b92d1269d6b63da2edc830def52ba433c7b727cc65c9bacR60-R63

File: iac/terraform/aws/cognito/cognito.tf (hunk 3 above)
Comment (Biplob14) [reply iac/terraform/aws/cognito/cognito.tf:89]: We don't need for now actually.

Comment (KhirulIslam) [reply iac/terraform/aws/cognito/cognito.tf:89]: remove that then, this can potentially create a bug later on

Comment (Biplob14) [reply iac/terraform/aws/cognito/cognito.tf]: removed from both places

**Review by KhirulIslam — APPROVED** [review]

**Review by sakibstark11 — APPROVED** [review]

**Pull request merged**